### PR TITLE
Install now-required `packaging` dependency of `asv` for benchmarks job in GitHub Actions

### PR DESCRIPTION
**Pull request by brocksam**
#### References to other Issues or PRs

Provides a temporary solution to https://github.com/sympy/sympy/issues/25561

#### Brief description of what is fixed or changed

asv=0.6.0 (released 20th August 2023) is causing a failure in our CI. Looks like a bug. This PR provides a temporary solution by pinning asv to the latest working version (asv==0.5.1), which we can keep until asv fix the issue in a new minor patch release.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by moorepants** [comment]
LGTM, I've opened issues on SymPy and ASV to see if we can figure out the correct long term fix. Enabling automerge.

**Comment by oscarbenjamin** [comment]
> Enabling automerge.

I've disabled automerge and instead changed the branch protection rules in the repo settings so that the benchmarks job is not "required" any more. That should mean that any PR can be merged even if the benchmarks job fails.

I don't object to merging this but possibly just making the job not required is better than pinning the version. If we pin the version then later we will have to unpin the version etc. If we don't pin the version then if asv releases an upstream fix then the CI job will just start working again.

On the other hand not pinning the version means that the benchmarks job fails in CI so we won't see the benchmark results and it might be confusing for some contributors to see that the job fails.

There are pros and cons either way but at least this problem does not block CI any more so we don't need to rush to merge it just for that.

**Comment by oscarbenjamin** [comment]
> the benchmarks job is not "required" any more

In general given that things can be flakey in CI (without systematic version pinning) I think that it is better to keep the list of "required" jobs as small as possible. I intend to leave this as "not required" rather than changing it back once the issue is fixed.

**Comment by brocksam** [comment]
> In general given that things can be flakey in CI (without systematic version pinning) I think that it is better to keep the list of "required" jobs as small as possible. I intend to leave this as "not required" rather than changing it back once the issue is fixed.

Understood. I think that's sound reasoning. And as you say above, we have a solution in place. Happy for this to be closed. Thanks @oscarbenjamin

**Comment by moorepants** [comment]
I'm good with that too.

**Comment by oscarbenjamin** [comment]
Let's keep this open for now. If the upstream issue does not get fixed then this can be merged if the benchmarks job is considered to be causing problems.

**Comment by asmeurer** [comment]
Yes, that's really annoying. There are feature requests for GitHub to fix this https://github.com/actions/runner/issues/2347, https://github.com/orgs/community/discussions/15452), but they don't seem to care. This is something I really miss about Travis CI.

File: .github/workflows/runtests.yml [hunk 1]
```diff
@@ -431,7 +431,7 @@ jobs:
       - uses: actions/setup-python@v4
         with:
           python-version: '3.11'
-      - run: pip install asv virtualenv
+      - run: pip install 'asv==0.5.1' virtualenv
```
Comment (HaoZeke) [inline .github/workflows/runtests.yml:434]:
```suggestion
      - run: pip install virtualenv packaging asv
```

Comment (HaoZeke) [reply .github/workflows/runtests.yml:434]: This should work.

**Review by HaoZeke — CHANGES_REQUESTED** [review]
The issue isn't `mamba`, it's that the `virtualenv` plugin grew a dependency on `packaging`.

File: .github/workflows/runtests.yml (hunk 1 above)
Comment (brocksam) [reply .github/workflows/runtests.yml]: Thanks for the hint. [This comment](https://github.com/airspeed-velocity/asv/pull/1324#issuecomment-1687429941) suggest we might be able to use plugin syntax so I'm trying that first as that'd be more robust.

Comment (HaoZeke) [reply .github/workflows/runtests.yml]: Unfortunately that's not a supported group yet. Though it should be present in 0.6.1

**Comment by brocksam** [comment]
This now solves the issue of the benchmarks job not running at all, but is showing the we're getting multiple failing benchmarks:

- The `cse.KalmanFilterMatrixEquationCSE.time_cse` benchmark is failing.
- We're getting multiple error messages related to `ImportError: Cannot import 'numpy' with 'import numpy; from numpy import *; from numpy.linalg import *' command`.

**Comment by HaoZeke** [comment]
> This now solves the issue of the benchmarks job not running at all, but is showing the we're getting multiple failing benchmarks:
> 
> * The `cse.KalmanFilterMatrixEquationCSE.time_cse` benchmark is failing.
> * We're getting multiple error messages related to `ImportError: Cannot import 'numpy' with 'import numpy; from numpy import *; from numpy.linalg import *' command`.

Not at a computer to check, but OTOH, this might because earlier `asv` versions were simply swallowing failing benchmarks (`scipy` also had to rework their benchmark suite which had silently broken). Were these benchmarks working before, locally? Does the `asv` configuration have `numpy` defined as a dependency? If it is only an optional dependency of `sympy` for example, it won't be installed into the environment by default.

**Comment by brocksam** [comment]
> earlier `asv` versions were simply swallowing failing benchmarks (`scipy` also had to rework their benchmark suite which had silently broken).

This is really useful information, thanks @HaoZeke. Almost certainly the failing benchmark was failing before asv 0.6.0.

> Does the `asv` configuration have `numpy` defined as a dependency? If it is only an optional dependency of `sympy` for example, it won't be installed into the environment by default.

Here's where my negligence with asv shows. [This is the environment that we're using in CI and it does reference NumPy](https://github.com/sympy/sympy_benchmarks/blob/eec71049a6020cc25a486535c44da2982ca92eb5/asv.conf.actions.json#L52-L56)

```
"matrix": {
        "fastcache": [],
        "mpmath": [],
        "numpy": []
},
```

mpmath is a required dependency. NumPy is an optional dependency.

**Comment by HaoZeke** [comment]
> > earlier `asv` versions were simply swallowing failing benchmarks (`scipy` also had to rework their benchmark suite which had silently broken).
> 
> This is really useful information, thanks @HaoZeke. Almost certainly the failing benchmark was failing before asv 0.6.0.
> 
> > Does the `asv` configuration have `numpy` defined as a dependency? If it is only an optional dependency of `sympy` for example, it won't be installed into the environment by default.
> 
> Here's where my negligence with asv shows. [This is the environment that we're using in CI and it does reference NumPy](https://github.com/sympy/sympy_benchmarks/blob/eec71049a6020cc25a486535c44da2982ca92eb5/asv.conf.actions.json#L52-L56)
> 
> ```
> "matrix": {
>         "fastcache": [],
>         "mpmath": [],
>         "numpy": []
> },
> ```
> 
> mpmath is a required dependency. NumPy is an optional dependency.

The environment looks alright, but I notice `numpy` isn't [imported in the benchmark itself](https://github.com/sympy/sympy_benchmarks/blob/eec71049a6020cc25a486535c44da2982ca92eb5/benchmarks/lambdify.py#L13-L33)?

**Comment by brocksam** [comment]
> The environment looks alright, but I notice `numpy` isn't [imported in the benchmark itself](https://github.com/sympy/sympy_benchmarks/blob/eec71049a6020cc25a486535c44da2982ca92eb5/benchmarks/lambdify.py#L13-L33)?

That's expected. The function that is being benchmarked here dynamically import the packages that it needs. I think this needs some further investigation by us on the sympy/sympy_benchmarks repo to ascertain whether this is a quirky bug with our `lambdify` function's importing or if it's something to do with asv not making NumPy available.

**Comment by oscarbenjamin** [comment]
I think the problem is that asv does not install numpy into the environment it creates and I'm pretty sure that's an asv bug.

Here `asv` says that it creates an environment called `virtualenv-py3.8-fastcache-mpmath-numpy` but whatever environment the benchmark runs in does not have numpy installed.

Steps to reproduce:
```console
$ python -m venv venv
$ source venv/bin/activate
$ pip install asv virtualenv packaging
...
Successfully installed asv-0.6.0 asv-runner-0.0.9 distlib-0.3.7 filelock-3.12.3 json5-0.9.14 packaging-23.1 platformdirs-3.10.0 pympler-1.0.1 pyyaml-6.0.1 tabulate-0.9.0 typing-extensions-4.7.1 virtualenv-20.24.4
$ git clone https://github.com/sympy/sympy_benchmarks.git
Cloning into 'sympy_benchmarks'...
...
$ cd sympy_benchmarks/
$ asv run --bench time_lambdify_evaluate
Couldn't load asv.plugins.mamba because
No module named 'mamba'
· Cloning project.
· Fetching recent changes.
· Creating environments....
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.8-fastcache-mpmath-numpy.
·· Building 46046153 <master> for virtualenv-py3.8-fastcache-mpmath-numpy..
·· Installing 46046153 <master> into virtualenv-py3.8-fastcache-mpmath-numpy..
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[ 0.00%] · For sympy commit 46046153 <master>:
[ 0.00%] ·· Benchmarking virtualenv-py3.8-fastcache-mpmath-numpy
[50.00%] ··· Running (lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate--).
[100.00%] ··· lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate                         1/2 failed
[100.00%] ··· =========== ============
                 param1               
              ----------- ------------
                ['math']   12.6±0.1μs 
               ['numpy']     failed   
              =========== ============
              For parameters: ['numpy']
              Traceback (most recent call last):
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/server.py", line 179, in _run_server
                  _run(run_args)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/run.py", line 65, in _run
                  skip = benchmark.do_setup()
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/time.py", line 80, in do_setup
                  result = Benchmark.do_setup(self)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/_base.py", line 632, in do_setup
                  setup(*self._current_params)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/benchmarks/lambdify.py", line 29, in setup
                  self.lmb = sp.lambdify(*_get_args_exprs(), modules=modules)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 796, in lambdify
                  buf = _get_namespace(m)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 946, in _get_namespace
                  _import(m)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 159, in _import
                  raise ImportError(
              ImportError: Cannot import 'numpy' with 'import numpy; from numpy import *; from numpy.linalg import *' command
```

**Comment by oscarbenjamin** [comment]
> I think the problem is that asv does not install numpy into the environment it creates and I'm pretty sure that's an asv bug.

I think for now that it is best to pin the asv version.

File: .github/workflows/runtests.yml [hunk 2]
```diff
@@ -431,7 +431,7 @@ jobs:
       - uses: actions/setup-python@v4
         with:
           python-version: '3.11'
-      - run: pip install asv virtualenv
+      - run: pip install asv virtualenv packaging
```
Comment (oscarbenjamin) [inline .github/workflows/runtests.yml:434]:
```suggestion
      - run: pip install 'asv==0.5.1' virtualenv packaging
```

**Comment by HaoZeke** [comment]
> I think the problem is that asv does not install numpy into the environment it creates and I'm pretty sure that's an asv bug.
> 
> Here `asv` says that it creates an environment called `virtualenv-py3.8-fastcache-mpmath-numpy` but whatever environment the benchmark runs in does not have numpy installed.
> 
> Steps to reproduce:
> 
> [SHELL-SESSION (49 lines)](#)
> ```
> $ python -m venv venv
> $ source venv/bin/activate
> $ pip install asv virtualenv packaging
> ...
> Successfully installed asv-0.6.0 asv-runner-0.0.9 distlib-0.3.7 filelock-3.12.3 json5-0.9.14 packaging-23.1 platformdirs-3.10.0 pympler-1.0.1 pyyaml-6.0.1 tabulate-0.9.0 typing-extensions-4.7.1 virtualenv-20.24.4
> $ git clone https://github.com/sympy/sympy_benchmarks.git
> Cloning into 'sympy_benchmarks'...
> ...
> $ cd sympy_benchmarks/
> $ asv run --bench time_lambdify_evaluate
> Couldn't load asv.plugins.mamba because
> No module named 'mamba'
> · Cloning project.
> · Fetching recent changes.
> · Creating environments....
> · Discovering benchmarks
> ·· Uninstalling from virtualenv-py3.8-fastcache-mpmath-numpy.
> ·· Building 46046153 <master> for virtualenv-py3.8-fastcache-mpmath-numpy..
> ·· Installing 46046153 <master> into virtualenv-py3.8-fastcache-mpmath-numpy..
> · Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
> [ 0.00%] · For sympy commit 46046153 <master>:
> [ 0.00%] ·· Benchmarking virtualenv-py3.8-fastcache-mpmath-numpy
> [50.00%] ··· Running (lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate--).
> [100.00%] ··· lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate                         1/2 failed
> [100.00%] ··· =========== ============
>                  param1               
>               ----------- ------------
>                 ['math']   12.6±0.1μs 
>                ['numpy']     failed   
>               =========== ============
>               For parameters: ['numpy']
>               Traceback (most recent call last):
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/server.py", line 179, in _run_server
>                   _run(run_args)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/run.py", line 65, in _run
>                   skip = benchmark.do_setup()
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/time.py", line 80, in do_setup
>                   result = Benchmark.do_setup(self)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/_base.py", line 632, in do_setup
>                   setup(*self._current_params)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/benchmarks/lambdify.py", line 29, in setup
>                   self.lmb = sp.lambdify(*_get_args_exprs(), modules=modules)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 796, in lambdify
>                   buf = _get_namespace(m)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 946, in _get_namespace
>                   _import(m)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 159, in _import
>                   raise ImportError(
>               ImportError: Cannot import 'numpy' with 'import numpy; from numpy import *; from numpy.linalg import *' command
> ```

`asv` has tests with `numpy`, so this is likely not the reason. Additionally, local testing of this:
```python
import numpy as np
from asv_runner.benchmarks.mark import parameterize_func_with

def time_sort(n):
    np.sort(np.random.rand(n))
time_sort.params = ([10, 100])
time_sort.param_names = ['n']

@parameterize_func_with({"n":[10, 100]})
def time_sort(n):
    np.sort(np.random.rand(n))

def time_ranges_multi(n, func_name):

    f = {'range': range, 'arange': np.arange}[func_name]
    for i in f(n):
        pass
time_ranges_multi.params = ([10, 100], ['range', 'arange'])
time_ranges_multi.param_names = ['n', 'function']

@parameterize_func_with({'n': [10, 100], 'func_name': ['range', 'arange']})
def time_ranges_multi(n, func_name):
    f = {'range': range, 'arange': np.arange}[func_name]
    for i in f(n):
        pass
```

With a config including:
```json
    "matrix": {
        "req": {
            "numpy": [""],
        },
    },
```

Works. If there's a reproducer though, please open an issue on the `asv` repo.

**Comment by oscarbenjamin** [comment]
> > Here `asv` says that it creates an environment called `virtualenv-py3.8-fastcache-mpmath-numpy` but whatever environment the benchmark runs in does not have numpy installed.
>
> `asv` has tests with `numpy`, so this is likely not the reason

I think it is the reason. Steps as above but let's use `asv==0.5.1` and see what is in the environment:
```console
$ pip install -q asv==0.5.1
$ asv run --bench time_lambdify_evaluate
· Fetching recent changes.
· Creating environments......
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.8-fastcache-mpmath-numpy.
·· Building 8f20ba58 <master> for virtualenv-py3.8-fastcache-mpmath-numpy..
·· Installing 8f20ba58 <master> into virtualenv-py3.8-fastcache-mpmath-numpy..
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] · For sympy commit 8f20ba58 <master>:
[  0.00%] ·· Benchmarking virtualenv-py3.8-fastcache-mpmath-numpy
[ 50.00%] ··· Running (lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate--).
[100.00%] ··· lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate                                 ok
[100.00%] ··· =========== =============
                 param1                
              ----------- -------------
                ['math']   12.4±0.04μs 
               ['numpy']    122±0.5μs  
              =========== =============

$ ls env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/
_distutils_hack            numpy.libs                    sympy
distutils-precedence.pth   pip                           sympy-1.13.dev0.dist-info
fastcache                  pip-23.2.1.dist-info          _virtualenv.pth
fastcache-1.1.0.dist-info  pip-23.2.1.virtualenv         _virtualenv.py
isympy.py                  pkg_resources                 wheel
mpmath                     __pycache__                   wheel-0.41.2.dist-info
mpmath-1.3.0.dist-info     setuptools                    wheel-0.41.2.virtualenv
numpy                      setuptools-68.1.2.dist-info
numpy-1.24.4.dist-info     setuptools-68.1.2.virtualenv
```
We can see that numpy is in the environment and that the "numpy" benchmark succeeds in 122us.

Now the same with `asv==0.6.0`:
```console
$ rm -r env # delete the environment created by asv 0.5.1
$ pip install -q asv==0.6.0
$ asv run --bench time_lambdify_evaluate
Couldn't load asv.plugins.mamba because
No module named 'mamba'
· Fetching recent changes.
· Creating environments...
· Discovering benchmarks
·· Uninstalling from virtualenv-py3.8-fastcache-mpmath-numpy.
·· Building 8f20ba58 <master> for virtualenv-py3.8-fastcache-mpmath-numpy..
·· Installing 8f20ba58 <master> into virtualenv-py3.8-fastcache-mpmath-numpy..
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[ 0.00%] · For sympy commit 8f20ba58 <master>:
[ 0.00%] ·· Benchmarking virtualenv-py3.8-fastcache-mpmath-numpy
[50.00%] ··· Running (lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate--).
[100.00%] ··· lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate                         1/2 failed
[100.00%] ··· =========== =============
                 param1                
              ----------- -------------
                ['math']   12.7±0.05μs 
               ['numpy']      failed   
              =========== =============
              For parameters: ['numpy']
              Traceback (most recent call last):
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/server.py", line 179, in _run_server
                  _run(run_args)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/run.py", line 65, in _run
                  skip = benchmark.do_setup()
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/time.py", line 80, in do_setup
                  result = Benchmark.do_setup(self)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/_base.py", line 632, in do_setup
                  setup(*self._current_params)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/benchmarks/lambdify.py", line 29, in setup
                  self.lmb = sp.lambdify(*_get_args_exprs(), modules=modules)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 796, in lambdify
                  buf = _get_namespace(m)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 946, in _get_namespace
                  _import(m)
                File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 159, in _import
                  raise ImportError(
              ImportError: Cannot import 'numpy' with 'import numpy; from numpy import *; from numpy.linalg import *' command

$ ls env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/
asv_runner                  pip-23.2.1.dist-info         setuptools-68.1.2.virtualenv
asv_runner-0.0.9.dist-info  pip-23.2.1.virtualenv        sympy
_distutils_hack             pkg_resources                sympy-1.13.dev0.dist-info
distutils-precedence.pth    __pycache__                  _virtualenv.pth
isympy.py                   pympler                      _virtualenv.py
mpmath                      Pympler-1.0.1.dist-info      wheel
mpmath-1.3.0.dist-info      setuptools                   wheel-0.41.2.dist-info
pip                         setuptools-68.1.2.dist-info  wheel-0.41.2.virtualenv
```
So the benchmark fails because numpy cannot be imported but that is not surprising because numpy is not in the environment.

There is no changelog for asv 0.6.0 so I don't know how to check whether anything has changed that would relevant for this:
https://asv.readthedocs.io/en/stable/changelog.html

It is clear though that something has changed between asv 0.5.1 and asv 0.6.0 meaning that asv no longer installs numpy into the environment and that is what causes the sympy benchmarks to fail with asv 0.6.0.

I tried changes like this but nothing seemed to help:
```diff
diff --git a/asv.conf.json b/asv.conf.json
index 9fcda26..8be1bd3 100644
--- a/asv.conf.json
+++ b/asv.conf.json
@@ -43,9 +43,11 @@
     // list indicates to just test against the default (latest)
     // version.
     "matrix": {
-        "fastcache": [],
-        "mpmath": [],
-        "numpy": [],
+        "req": {
+            "fastcache": [""],
+            "mpmath": [""],
+            "numpy": [""],
+        },
     },
 
     // The directory (relative to the current directory) that benchmarks are
```

**Comment by HaoZeke** [comment]
Sorry

> > > Here `asv` says that it creates an environment called `virtualenv-py3.8-fastcache-mpmath-numpy` but whatever environment the benchmark runs in does not have numpy installed.
> > 
> > 
> > `asv` has tests with `numpy`, so this is likely not the reason
> 
> I think it is the reason. Steps as above but let's use `asv==0.5.1` and see what is in the environment:
> 
> ```
> $ pip install -q asv==0.5.1
> $ asv run --bench time_lambdify_evaluate
> · Fetching recent changes.
> · Creating environments......
> · Discovering benchmarks
> ·· Uninstalling from virtualenv-py3.8-fastcache-mpmath-numpy.
> ·· Building 8f20ba58 <master> for virtualenv-py3.8-fastcache-mpmath-numpy..
> ·· Installing 8f20ba58 <master> into virtualenv-py3.8-fastcache-mpmath-numpy..
> · Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
> [  0.00%] · For sympy commit 8f20ba58 <master>:
> [  0.00%] ·· Benchmarking virtualenv-py3.8-fastcache-mpmath-numpy
> [ 50.00%] ··· Running (lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate--).
> [100.00%] ··· lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate                                 ok
> [100.00%] ··· =========== =============
>                  param1                
>               ----------- -------------
>                 ['math']   12.4±0.04μs 
>                ['numpy']    122±0.5μs  
>               =========== =============
> 
> $ ls env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/
> _distutils_hack            numpy.libs                    sympy
> distutils-precedence.pth   pip                           sympy-1.13.dev0.dist-info
> fastcache                  pip-23.2.1.dist-info          _virtualenv.pth
> fastcache-1.1.0.dist-info  pip-23.2.1.virtualenv         _virtualenv.py
> isympy.py                  pkg_resources                 wheel
> mpmath                     __pycache__                   wheel-0.41.2.dist-info
> mpmath-1.3.0.dist-info     setuptools                    wheel-0.41.2.virtualenv
> numpy                      setuptools-68.1.2.dist-info
> numpy-1.24.4.dist-info     setuptools-68.1.2.virtualenv
> ```
> 
> We can see that numpy is in the environment and that the "numpy" benchmark succeeds in 122us.
> 
> Now the same with `asv==0.6.0`:
> 
> ```
> $ rm -r env # delete the environment created by asv 0.5.1
> $ pip install -q asv==0.6.0
> $ asv run --bench time_lambdify_evaluate
> Couldn't load asv.plugins.mamba because
> No module named 'mamba'
> · Fetching recent changes.
> · Creating environments...
> · Discovering benchmarks
> ·· Uninstalling from virtualenv-py3.8-fastcache-mpmath-numpy.
> ·· Building 8f20ba58 <master> for virtualenv-py3.8-fastcache-mpmath-numpy..
> ·· Installing 8f20ba58 <master> into virtualenv-py3.8-fastcache-mpmath-numpy..
> · Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
> [ 0.00%] · For sympy commit 8f20ba58 <master>:
> [ 0.00%] ·· Benchmarking virtualenv-py3.8-fastcache-mpmath-numpy
> [50.00%] ··· Running (lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate--).
> [100.00%] ··· lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate                         1/2 failed
> [100.00%] ··· =========== =============
>                  param1                
>               ----------- -------------
>                 ['math']   12.7±0.05μs 
>                ['numpy']      failed   
>               =========== =============
>               For parameters: ['numpy']
>               Traceback (most recent call last):
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/server.py", line 179, in _run_server
>                   _run(run_args)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/run.py", line 65, in _run
>                   skip = benchmark.do_setup()
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/time.py", line 80, in do_setup
>                   result = Benchmark.do_setup(self)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/asv_runner/benchmarks/_base.py", line 632, in do_setup
>                   setup(*self._current_params)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/benchmarks/lambdify.py", line 29, in setup
>                   self.lmb = sp.lambdify(*_get_args_exprs(), modules=modules)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 796, in lambdify
>                   buf = _get_namespace(m)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 946, in _get_namespace
>                   _import(m)
>                 File "/home/oscar/current/active/tmp/bench/sympy_benchmarks/env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/sympy/utilities/lambdify.py", line 159, in _import
>                   raise ImportError(
>               ImportError: Cannot import 'numpy' with 'import numpy; from numpy import *; from numpy.linalg import *' command
> 
> $ ls env/1917629413157ce7eb6bf70c1499b87b/lib/python3.8/site-packages/
> asv_runner                  pip-23.2.1.dist-info         setuptools-68.1.2.virtualenv
> asv_runner-0.0.9.dist-info  pip-23.2.1.virtualenv        sympy
> _distutils_hack             pkg_resources                sympy-1.13.dev0.dist-info
> distutils-precedence.pth    __pycache__                  _virtualenv.pth
> isympy.py                   pympler                      _virtualenv.py
> mpmath                      Pympler-1.0.1.dist-info      wheel
> mpmath-1.3.0.dist-info      setuptools                   wheel-0.41.2.dist-info
> pip                         setuptools-68.1.2.dist-info  wheel-0.41.2.virtualenv
> ```
> 
> So the benchmark fails because numpy cannot be imported but that is not surprising because numpy is not in the environment.
> 
> There is no changelog for asv 0.6.0 so I don't know how to check whether anything has changed that would relevant for this: https://asv.readthedocs.io/en/stable/changelog.html
> 
> It is clear though that something has changed between asv 0.5.1 and asv 0.6.0 meaning that asv no longer installs numpy into the environment and that is what causes the sympy benchmarks to fail with asv 0.6.0.
> 
> I tried changes like this but nothing seemed to help:
> 
> ```diff
> diff --git a/asv.conf.json b/asv.conf.json
> index 9fcda26..8be1bd3 100644
> --- a/asv.conf.json
> +++ b/asv.conf.json
> @@ -43,9 +43,11 @@
>      // list indicates to just test against the default (latest)
>      // version.
>      "matrix": {
> -        "fastcache": [],
> -        "mpmath": [],
> -        "numpy": [],
> +        "req": {
> +            "fastcache": [""],
> +            "mpmath": [""],
> +            "numpy": [""],
> +        },
>      },
>  
>      // The directory (relative to the current directory) that benchmarks are
> ```

Sorry about this, turns out (will be fixed in the next release) that with `virtualenv` in the current release `pip+` is required for `pip` dependencies. That is:

```json
        "req": {
            "pip+fastcache": [""],
            "pip+mpmath": [""],
            "pip+numpy": [""],
        },
```

This is a breaking change and should have been documented.

**Comment by oscarbenjamin** [comment]
> turns out that with `virtualenv` in the current release `pip+` is required for `pip` dependencies.

Thanks!

I've opened #25625

**Comment by oscarbenjamin** [comment]
> > turns out that with `virtualenv` in the current release `pip+` is required for `pip` dependencies.
> 
> Thanks!
> 
> I've opened #25625

Yes, that fixed it. Thanks again!

In https://github.com/sympy/sympy_benchmarks/pull/98 I added `pip+` to most of the `asv.*.json` files but not to the ones using conda since I presume it is not supposed to be used there but I haven't tested them. If conda is used should it be `conda+`?
https://github.com/sympy/sympy_benchmarks/blob/207b878a62d2126de81b913ff587192750ab0c12/asv.conf.conda.json#L32-L49

**Comment by HaoZeke** [comment]
Excellent. The one with conda should be fine as is. Actually asv should always default to the environment manager (conda /mamba /pip for virtualenv) and the only one which needs to be set explicitly is pip , so its an asv bug which will be fixed in the upcoming release.

**Pull request merged**